generation.go

```diff
@@ -0,0 +1,386 @@
+package gen
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+)
+
+type Generation struct {
+	api *API
+
+	enums     []*Enum
+	functions []*Function
+	structs   []*Struct
+
+	Lookup
+}
+
+func NewGeneration(a *API) *Generation {
+	gen := &Generation{
+		api: a,
+
+		Lookup: Lookup{
+			lookupEnum:        map[string]*Enum{},
+			lookupNonTypedefs: map[string]string{},
+			lookupStruct: map[string]*Struct{
+				"cxstring": &Struct{
+					Name:  "cxstring",
+					CName: "CXString",
+				},
+			},
+		},
+	}
+
+	return gen
+}
+
+func (g *Generation) AddHeaderFiles(headerFiles []*HeaderFile) {
+	for _, h := range headerFiles {
+		for _, e := range h.Enums {
+			g.enums = append(g.enums, e)
+			g.RegisterEnum(e)
+		}
+
+		for _, s := range h.Structs {
+			g.structs = append(g.structs, s)
+			g.RegisterStruct(s)
+		}
+
+		for _, f := range h.Functions {
+			g.functions = append(g.functions, f)
+		}
+	}
+}
+
+func (g *Generation) Generate() error {
+	// Prepare all functions
+	clangFile := newFile("clang")
+
+	for _, f := range g.functions {
+		var fname string
+		if g.api.PrepareFunctionName != nil {
+			fname = g.api.PrepareFunctionName(g, f)
+			f.Name = fname
+		} else {
+			fname = f.Name
```

Comment:
Let's init fname with f.Name instead of doing this in the else block. I know that it could happen that the assignment is overwritten, but it is much easier to read and we do not program for performance atm.